torch/lib/c10d/ProcessGroup.cpp

```diff
@@ -77,10 +78,18 @@ ProcessGroup::Work::Work(
         }
       }
       recordingFunction->before(profilingTitle, inputs);
-      std::function<void()> end_handler = [this, recordingFunction]() {
+      std::function<void()> end_handler = [recordingFunction]() {
         recordingFunction->end();
       };
-      recordFunctionEndCallback_ = at::wrapPropagateTLSState(end_handler);
+      auto recordFunctionEndCallback = at::wrapPropagateTLSState(end_handler);
+
+      // Add a callback that runs profiling end callbacks. wrapCallback() in
+      // CUDA future blocks the stream this callback runs on the corresponding
+      // cudaEvents_ ensuring appropriate synchronization.
+      future_->addCallback(
+          [recordFunctionEndCallback](c10::ivalue::Future& /*unused*/) {
```

Comment:
Consider `std::move`ing `recordFunctionEndCallback` into the lambda. It could save a copy. You could also avoid the temporary variable entirely:
```suggestion
          [recordFunctionEndCallback=at::wrapPropagateTLSState(end_handler)](c10::ivalue::Future& /*unused*/) {
```